src/jit/importer.cpp

```diff
@@ -851,14 +852,25 @@ GenTreeArgList* Compiler::impPopList(unsigned count, CORINFO_SIG_INFO* sig, GenT
             // Morph trees that aren't already OBJs or MKREFANY to be OBJs
             assert(ti.IsType(TI_STRUCT));
             structType = ti.GetClassHandleForValueClass();
+
+            bool forceNormalization = false;
+            if (varTypeIsSIMD(temp))
+            {
+                // We need to ensure that fgMorphArgs will use the correct struct handle to ensure proper
+                // ABI handling of this argument.
+                if (gtGetStructHandleIfPresent(temp) != structType)
```

Comment:
This happens, for example, when we have a SIMD intrinsic that takes one type and returns another (e.g. takes a SIMD with a float base type and returns one with an int). We currently don't examine all the signatures of all the methods etc. that we encounter. I started down the path of asserting here (and fixing all the places where we don't capture the handle), but decided that was best left for a future PR.
I'd really like to eventually get rid of much of the handle-handling for "known" types, but it's a bit complicated for SIMD types because of the ambiguity between `Vector4`, `Vector<T>` on an SSE system, and `Vector128<T>`. The first is a non-opaque type (a "normal" struct, e.g. for ABI purposes), while the other two are opaque vector types.

Reply:
`Vector4` is officially a struct of 4 floats. So, when we begin to support vector ABIs (which I'm attempting to do for Arm64 currently), it *won't* be passed in a vector register.

Reply:
> So, when we begin to support vector ABIs (which I'm attempting to do for Arm64 currently), it won't be passed in a vector register.

Wouldn't it still be passed in a vector register for certain ABIs? Some, like the `System V ABI` (or `__vectorcall` on Windows) have special logic for structs like `Vector4` or `Matrix4x4` that allows them to be passed in register.

Reply:
AFAIR they do go in registers but in a rather peculiar manner - X and Y in one XMM reg and Z and W in another XMM reg. And `__vectorcall` places each component in a separate XMM register, it doesn't put the entire `Vector4` in a single register.

Which is actually rather unfortunate. It may be better to actually treat `Vector2/3/4` specially, as if they are `__m128`. Not only that it would be more efficient but it also avoids JIT complications such as the one you described. The fact that they are structs containing a couple of float fields doesn't mean that they must be treated as normal structs. In Visual C++ `__m128` is actually a giant union of structs after all.

Reply:
> It may be better to actually treat Vector2/3/4 specially, as if they are __m128

The problem with that is that these are specifically designed to be interop types, so we need to pass them compatibly.

Reply:
It's easy enough to validate with an existing compiler: https://godbolt.org/z/t35LGh
* `__m128` is the ABI type
* `V4` is a struct containing 4 floats
* `MatrixM128` is a struct containing 4 `__m128`
* `MatrixV4` is a struct containing 4 `V4`

The results are:
* On Windows, with the default calling convention, nothing is passed in register (neither `__m128`, `V4`, `MatrixM128`, nor `MatrixV4`)
* On Windows, with `__vectorcall`, `__m128`, `V4`, and `MatrixM128` are passed in register (while registers are available). `MatrixV4` is passed via Memory
  * This is due to the HVA (Homogeneous Vector Aggregate) logic
* On Unix, `__m128` and `V4` are passed in register  (while registers are available). `MatrixM128` and `MatrixV4` are passed in Memory

It is worth noting that, for Unix:
> The post merger clean up described later ensures that, for the processors that do not support the __m256 type, if the size of an object is larger than two eightbytes and the first eightbyte is not SSE or any other eightbyte is not SSEUP, it still has class MEMORY. This in turn ensures that for processors that do support the __m256 type, if the size of an object is four eightbytes and the first eightbyte is SSE and all other eightbytes are SSEUP, it can be passed in a register. This also applies to the __m512 type. That is for processors that support the __m512 type, if the size of an object is eight eightbytes and the first eightbyte is SSE and all other eightbytes are SSEUP, it can be passed in a register, otherwise, it will be passed in memory.

So, some of the logic changes if the compiler is emitting AVX or AVX512 compatible code.

Reply:
Hmm, I thought that this could be done for `__vectorcall` but it seems like it's not possible.

In the standard Windows calling convention, `__m128` is passed as a struct so if we make `Vector4` behave like `__m128` then interop should continue to work. `__m128` and `Vector4` would be passed in XMM registers only when `__vectorcall` is used.

But then it turns out that only `__m128` arguments are treated as normal structs in the standard Windows calling convention, `__m128` **returns** are not treated as normal structs. Sheesh. ABIs are a nightmare. And unlike native compilers the JIT doesn't have the luxury to use custom calling conventions.

Reply:
Updated the notes above to more clearly call out when passing in `register` means a single vs multiple register and the behavior for returns